### PR TITLE
encryption: support KMS on Azure

### LINKED ISSUE #14693: encryption: make encryption at rest adaptive to Azure.
## Development Task
In previous works, TiKV already supports encryption at rest, and make it adaptive to AWS by AWS KMS.

So, it's necessary to make it adaptive to multi-clouds backends, typically like Azure, and maybe for GCP in the future.
And there are several tasks for it:
- ~~[ ] Make the previous management of master key adaptive to multi-clouds.~~
- [x] Supply Azure APIs to support encryption at rest on Azure.
- [ ] Support encryption at rest on Azure.

### PR DESCRIPTION
**Pull request by LykxSassinator**
### What is changed and how it works?

Issue Number: Close #14693 

What's Changed:

```commit-message
This pull request introduces the integration of the Azure Key Management Service (KMS) into the project. 
The KMS functionality allows for secure key management and encryption operations in the Azure environment.

And given the ongoing discussions regarding the final representation of encryption support on multi-backend clouds, this implementation aims to be as compatible as possible with previous works.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None.
```

### CODE REVIEW

File: components/cloud/azure/src/kms.rs [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+        // Priority: explicit certificate > path of local certificate > client secret.
+        // And credentials for accessing KeyVault and Hsm should be different.
+        if let Some(certificate) = azure_cfg.client_certificate.clone() {
+            // Certificate to accessing KeyVault.
+            let (keyvault_credential, hsm_credential) = (
+                ClientCertificateCredentialExt::new(
+                    azure_cfg.tenant_id.clone(),
+                    azure_cfg.client_id.clone(),
+                    certificate.clone(),
+                    "".to_owned(),
```
Comment (overvenus) [inline components/cloud/azure/src/kms.rs:93]: Empty `certificate_pass`?  What if it's protected by a password?

File: components/cloud/azure/src/kms.rs [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+        ENCRYPTION_VENDOR_NAME_AZURE_KMS
+    }
+
+    // On decrypt failure, the rule is to return WrongMasterKey error in case it is
+    // possible that a wrong master key has been used, or other error
+    // otherwise.
+    async fn decrypt_data_key(&self, data_key: &EncryptedKey) -> Result<Vec<u8>> {
+        let decrypt_params = DecryptParameters {
+            ciphertext: data_key.clone().into_inner(),
+            decrypt_parameters_encryption: CryptographParamtersEncryption::Rsa(
```
Comment (overvenus) [inline components/cloud/azure/src/kms.rs:159]: Could you add some comment about why we choose `EncryptionAlgorithm::RsaOaep256`?

File: components/encryption/export/examples/ecli.rs [hunk 3]
```diff
@@ -61,6 +62,27 @@ struct KmsCommand {
     /// Remote region.
     #[structopt(long)]
     region: Option<String>,
+    /// Azure KMS
+    #[structopt(subcommand)]
+    azure: Option<KmsSubCommandAzure>,
```
Comment (overvenus) [inline components/encryption/export/examples/ecli.rs:67]: How about renaming `Command::Kms` to `Command::Aws` and adding `Command::Azure(KmsSubCommandAzure)`?

File: components/cloud/azure/src/kms.rs (hunk 1 above)
Comment (LykxSassinator) [reply components/cloud/azure/src/kms.rs:93]: Got. I'll expose this configuration, I've missed this point.

**Review by overvenus — APPROVED** [review]

**Review by Leavrth — APPROVED** [review]

File: components/encryption/src/config.rs [hunk 4]
```diff
@@ -46,6 +47,31 @@ pub struct FileConfig {
...
+    pub hsm_url: String,
+    /// Authorized certificate, the certificate is expected to be in base64
+    /// encoded PKCS12 format
+    pub client_certificate: Option<String>,
+    /// Path of local authorized certificate
+    pub client_certificate_path: Option<String>,
+    /// Password for the certificate
+    pub client_certificate_password: String,
+    /// Secret of the client.
+    pub client_secret: Option<String>,
```
[the remark this answers is not in the document]
Comment (LykxSassinator) [reply components/encryption/src/config.rs:72]: Nope.
As the `Debug` function is override in `azure/kms.rs`:
image

The infos on client secrets, certificates and its related passwords won't be exposed to logs while logging.

[the remark this answers is not in the document]
Comment (LykxSassinator) [reply components/encryption/src/config.rs:72]: Thx for the reminder. Fixed.

**Review by 3pointer — APPROVED** [review]

**Pull request merged**